### DISCUSSION TITLE
Parsing a number that uses comma as the decimal separator

### DISCUSSION THREAD

**Discussion by oehnstro** [colinhacks/zod#3339] [Q&A] [answered]
Hi, I have a seemingly small task that I can't seem to get right.

I need to parse the response from an external API. One field in the response is a string that contains a number that uses comma `,` as the decimal separator. This means I can't coerce it to be a number directly. I'm trying to use `.transform()` and `.pipe()`, but for some reason the compiler can't see that the type ends up as a number and not a string. Below is an example.

```
// Example of trying to parse a string that uses comma as decimal separator into a number
// e.g. "9,99" -> 9.99
const example: z.ZodSchema<{
  decimalNumber: number;
}> = z.object({
  decimalNumber: z.string().transform((val) => Number(`${val}`.replace(",", "."))).pipe(z.number()),
});
```

The compiler then complains that:

```
Type 'ZodObject<{ decimalNumber: ZodPipeline<ZodEffects<ZodString, number, string>, ZodNumber>; }, "strip", ZodTypeAny, { ...; }, { ...; }>' is not assignable to type 'ZodType<{ decimalNumber: number; }, ZodTypeDef, { decimalNumber: number; }>'.
  The types of '_input.decimalNumber' are incompatible between these types.
    Type 'string' is not assignable to type 'number'.
```

What am I missing here?

Any help or pointers in the right direction would be greatly appreciated 😄

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const example = z.object( {
    decimalNumber: z.string().transform( ( val ) => Number( `${ val }`.replace( ",", "." ) ) ).pipe( z.number() ),
} ) satisfies z.ZodSchema<{ decimalNumber: number }, z.ZodTypeDef, { decimalNumber: string }>
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by oehnstro** [reply]
Thank you! Your code works as expected for the example I gave. 

In our case we are dealing with a recursive type in addition to this, so I wasn't able to use your solution. The actual code works perfectly, so we ended up casting the field as any to avoid the type issues and wrote some additional test to make sure the code keeps working.
